### PR TITLE
US134384 - Add new subscription controllers

### PR DESCRIPTION
**Pull request by svanherk**
Cleaned up version of https://github.com/BrightspaceUI/core/pull/1901.  Creates a new `controllers` folder with new subscriber controllers inside, along with tests and a README.  The README was surprisingly difficult to write and I don't love it - feedback welcome

### CODE REVIEW

File: controllers/subscriber/test/subscriberControllers.test.js [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+		}
+	}
+);
+
+const idSubscriber = defineCE(
+	class extends LitElement {
+		static get properties() {
+			return {
+				for: {
+					converter: {
```
Comment (svanherk) [inline controllers/subscriber/test/subscriberControllers.test.js:121]: So here's an example of the converter property to change it from a string of space-delimited ids.  Pretty clean, but means that any component that supports the list (versus just one id) would need to use this for it's `for` property.  Or we can just always expect a string at https://github.com/BrightspaceUI/core/pull/1927/files#diff-647765ef2f64b4feb979d54c882d976462c39916efc668bc36b1e6c85814cbbdR143, and do the string to array parsing there instead. Thoughts?

File: controllers/subscriber/subscriberControllers.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		if (this._subscriptionController) this._subscriptionController.hostDisconnected();
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+		if (this._subscriptionController) this._subscriptionController.hostUpdated(changedProperties);
+	}
+```
```
Comment (dlockhart) [inline controllers/subscriber/subscriberControllers.md:153]: I found the sample code a little hard to follow in terms of tying what the example was trying to do to the core concepts. 🐶  🐱  It's very cute though and I did learn fun facts about tongue prints!

Maybe a metaphor that also uses a subscription pattern could work? Like a "newspaper" <-> "reader" or a "cable company" <-> "viewer" type metaphor? Do people still even subscribe to newspapers?!

File: controllers/subscriber/test/subscriberControllers.test.js (hunk 1 above)
Comment (svanherk) [reply controllers/subscriber/test/subscriberControllers.test.js:121]: Yeah, or we can just always expect a string and do the array conversion for them in the controller.  That's _probably_ easier...

File: controllers/subscriber/subscriberControllers.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
+# Subscriber Controllers
```
[the remark this answers is not in the document]
Comment (svanherk) [reply controllers/subscriber/subscriberControllers.md:1]: Good call, did all the development with it in the `helpers` folder and followed the convention there

File: controllers/subscriber/test/subscriberControllers.test.js (hunk 1 above)
Comment (dlockhart) [reply controllers/subscriber/test/subscriberControllers.test.js:121]: I'm good with that 👍

File: controllers/subscriber/subscriberControllers.md (hunk 2 above)
Comment (svanherk) [reply controllers/subscriber/subscriberControllers.md:153]: Yeah I adapted this from the Provider/Requester README, but using a more intuitive subscription example makes sense.  I'll try it with cable channels and viewers

File: controllers/subscriber/test/subscriberControllers.test.js [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+
+			expect(registry._onSubscribeTargets.length).to.equal(2);
+			expect(registry._onSubscribeTargets[0]).to.equal(elem.querySelector('#id'));
+			expect(registry._onSubscribeTargets[1]).to.equal(elem.querySelector('#event'));
+		});
+
+		it('Additional subscribers can be subscribed manually', () => {
+			expect(registry._eventSubscribers.subscribers.size).to.equal(1);
+			expect(registry._idSubscribers.subscribers.size).to.equal(1);
+			expect(registry._onSubscribeTargets.length).to.equal(2);
```
[the remark this answers is not in the document]
Comment (svanherk) [reply controllers/subscriber/test/subscriberControllers.test.js:192]: Yeah, I like making the tests "readable" but it's definitely redundant.  Can cleanup a bit

File: controllers/subscriber/subscriberControllers.md (hunk 2 above)
Comment (svanherk) [reply controllers/subscriber/subscriberControllers.md]: Ok, updated - let me know if that works better. It's still a bit of a weird example when looking at the resulting html, but I wanted to show how event subscribers must go inside the registry component, and id subscribers can go outside.

Comment (dlockhart) [reply controllers/subscriber/subscriberControllers.md]: I like this a lot! 👍

**Review by dlockhart — APPROVED** [review]
This is looking great to me!
